### ISSUE TITLE
Shouln't https be on port 443?

### ISSUE THREAD

**Issue by dedani** [SergioBenitez/Rocket#652]
I notice that no matter if a server is launched (ROCKET_ENV=prod) with or without TLS support, it i hosted on port 80. But shouldn't a https server default to port 443?

**Comment by vhakulinen** [comment]
Yes. HTTPS is on port 443 by default.

**Comment by SergioBenitez** [comment]
The challenge here is as follows: Rocket can certainly set the port to 443 by default if TLS is being used. This conditional default mechanism, however, is not something that can be exposed to users for them to do themselves. That is, it wouldn't be possible for the users to set different defaults based on whether TLS is enabled or not. So, the question becomes: should Rocket do something by default that users are unable to do themselves?

**Comment by mettke** [comment]
I'm wondering whether we should use two separate servers. One for HTTP and one for HTTPS as described in [how-to-build-multiple-concurrent-servers-with-rust-and-tokio](https://stackoverflow.com/questions/42759882/how-to-build-multiple-concurrent-servers-with-rust-and-tokio). The Configuration would then contain `http_port = 80` and `https_port = 443` as default. The only problem would be when to start HTTP, https or both. 

One way is to use the current method and add a new flag like `serve_http = true`. It would provide the following situations with the unfortunate first situation where no server gets started.

|  TLS Configured | serve_http set  | ports binded |
|:-------------:|:-------------:|:-----:|
|   ✗  | ✗ | - |
|  ✗    | ✓      |  HTTP |
|  ✓    | ✗      |   HTTPS |
|  ✓    | ✓      |  HTTP, HTTPS |

Another way would be to have a flag like `serve_on = ["http"|"https"|"both"]` which would remove the situation where no server gets started. It, however, introduces a situation where we say `serve_one = https` and no certs configured thus resulting in an error state. The configuration would default to `serve_on = http` and requires the following changes to serve on https or both

```
certs = "/path/to/certs.pem"
key = "/path/to/key.pem"
serve_on = ["https"|"both"]
```

In my opinion, there is a reason for systems like httpd, nginx, tomcat, [...] to allow users to bind to both protocols. Sometimes HTTP is required and shouldn't require the application to be started a second time with a different configuration.

**Comment by NathanielAJohnson** [comment]
@mettke 's idea is great.  I always need a application level re-direct for sites in general specifically when i renew credentials via a webserver (Letsencrypt, oddly, requires an unencrypted verification).

**Comment by jebrosen** [comment]
This caught my attention:

> I notice that no matter if a server is launched (ROCKET_ENV=prod) with or without TLS support, **it i hosted on port 80**. But shouldn't a https server default to port 443?

I think this issue was opened before the 0.4 release, which included #450 (closing #317) changing the default port to 8000 in all environments. For this to be true with now the port would have to be set to 80 manually. In light of that change, I think this issue is less relevant.

---

That said, I see a few ways to possibly improve this:

* ~~If no port has been specified, launch on port 80 for HTTP and 443 for HTTPS.~~
  * This was already covered in #317, Rocket shouldn't be doing this for the environments it's most often used in today.
* Allow specification of "port=auto" to choose 80 or 443 depending on presence/absence of TLS
  * Pro: Easy access to the desired behavior.
  * Con: Non-obvious logic to implement and test, and I don't see much value in doing this over specifying "port=443" or "port=80" depending on whether TLS is being used or not.
* Support multiple listening addresses/ports, possibly with configuration parameters similar to what @mettke described.
  * Pro: Greatest amount of flexibility. This would be an even more important consideration when binding to Unix sockets is supported (like #594 implements).
  * Con: Most implementation effort and open questions.
